core/src/main/java/io/grpc/LoadBalancer.java

```diff
@@ -0,0 +1,98 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import io.grpc.internal.ClientTransport;
+
+import java.net.SocketAddress;
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A pluggable component that receives resolved addresses from {@link NameResolver} and provides the
+ * channel a usable transport when asked.
+ *
+ * <p>It owns the transports and manages their lifecycles.
+ */
+// TODO(zhangkun83): since it's also used for non-loadbalancing cases like pick-first,
+// "RequestRouter" might be a better name.
+@ExperimentalApi
+public interface LoadBalancer {
+  /**
+   * Start the load-balancing task for a service.
+   *
+   * @param serviceName the DNS-style service name, which is also the authority
+   * @param tm the interface where an {@code LoadBalancer} implementation gets connected transports
+   *           from
+   * @param executor used to run internal tasks
+   */
+  Handle start(String serviceName, TransportManager tm, Executor executor);
+
+  /**
+   * Interface to interact with a load-balancing task and for the task to receive updates from
+   * {@link NameResolver} and the channel.
+   */
+  @ThreadSafe
+  abstract class Handle {
+    /**
+     * Pick a transport that Channel will use for next RPC.
+     *
+     * @param requestKey for affinity-based routing
+     */
+    public abstract ListenableFuture<ClientTransport> pickTransport(
+        @Nullable RequestKey requestKey);
+
+    /**
+     * Stops the load-balancing task.
+     */
+    public abstract void shutdown();
+
+    /**
+     * Handles newly resolved addresses and service config from name resolution system.
+     */
+    public void handleResolvedAddresses(ResolvedServerInfo[] servers, Attributes config) { }
```

Comment:
Maybe use a collection/list instead of array to prevent unnecessary copy? Ditto elsewhere with []s.